### PR TITLE
Mark `logger_connector` as protected

### LINKED ISSUE #9778: Mark the trainer's connectors as protected
## Proposed refactoring or deprecation

Mark these attributes of the Trainer as protected:
https://github.com/PyTorchLightning/pytorch-lightning/blob/ab207921b9149a2c20ff9f797b6381f61cb09b1e/pytorch_lightning/trainer/trainer.py#L389-L419

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

By python coding conventions, the connectors are technically part of the trainer's public API, despite not being documented and not intending to be used by callers. 

However, we do not honor backward compatibility for these connectors because we (framework developers) consider them as implementation details of the trainer. Examples:
https://github.com/PyTorchLightning/pytorch-lightning/pull/9680
https://github.com/PyTorchLightning/pytorch-lightning/pull/9566

### Pitch
Let's make this explicit that these connectors are not intended to be called by end users
Proposal: prefix all the attributes with an underscore to mark them as protected. 
This gives us users more clarity around what is part of the public trainer API, and gives framework developers more flexibility to evolve the implementation here.

To make the PRs easy to review, we can migrate one connector at a time (1 PR per connector)

### Alternatives
Keep as is

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?

Fixes #9778 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/device_stats_monitor.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def on_train_batch_end(
         if not trainer.loggers:
             raise MisconfigurationException("Cannot use `DeviceStatsMonitor` callback with `Trainer(logger=False)`.")
 
-        if not trainer.logger_connector.should_update_logs:
+        if not trainer._logger_connector.should_update_logs:
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/device_stats_monitor.py:83]: n00b question: why does the callback need to check this vs simply calling logger.log_metrics?

**Review by ananthsub — APPROVED** [review]

**Review by krshrimali — APPROVED** [review]
Neat, LGTM!

Just checked to make sure that there are no other instances of `logger_connector` (to be replaced with `_logger_connector`), and there are None. For reviewers:

```bash
$ grep -wr "logger_connector" --exclude=\*.{txt,md} --exclude-dir=".github" | grep -v "from"
```

Outputs:
```python
tests/loops/test_evaluation_loop.py:    "pytorch_lightning.trainer.connectors.logger_connector.logger_connector.LoggerConnector.log_eval_end_metrics"
tests/trainer/logging_/test_loop_logging.py:        "pytorch_lightning.trainer.connectors.logger_connector.result._ResultCollection.log", return_value=None
pytorch_lightning/trainer/connectors/logger_connector/result.py:    :class:`~pytorch_lightning.trainer.connectors.logger_connector.result._ResultMetric`
pytorch_lightning/trainer/connectors/logger_connector/result.py:    Collection (dictionary) of :class:`~pytorch_lightning.trainer.connectors.logger_connector.result._ResultMetric` or
pytorch_lightning/trainer/connectors/logger_connector/result.py:    :class:`~pytorch_lightning.trainer.connectors.logger_connector.result._ResultMetricCollection`
```

**Review by kaushikb11 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/device_stats_monitor.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/device_stats_monitor.py:83]: to follow the `log_every_n_steps` criteria since this is not controlled by loggers.

**Review by awaelchli — APPROVED** [review]
